lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -111,12 +111,34 @@ varConformsToCodable(DeclContext *DC, VarDecl *varDecl, ProtocolDecl *proto) {
   // }
   //
   // struct Bar : Codable {
-  //   var x: Int // <- we get to valuate x's var decl here, but its type
+  //   var x: Int // <- we get to evaluate x's var decl here, but its type
   //              //    hasn't yet been evaluated
   // }
+
+  // If the variable type is the same as the context type, then we can end up
+  // with inconsistent witnesses. For example:
+  //
+  // class Foo: Decodable { // <- Evaluating Foo will trigger derivation
+  //                         // of init(from:) but we cannot overwrite the
+  //                         // (invalid) witness we set below.
+  //
+  //  let bar = Foo() // <- Evaluating bar will trigger derivation of
+  //                  // init(from:) but we cannot derive it yet.
+  // }
+  //
+  // So, if the variable type and context type are the same, then let's just
+  // return a failure because the variable does not conform to the protocol
+  // because the type itself doesn't.
+  auto varType = varDecl->getValueInterfaceType();
+  if (proto->getKnownDerivableProtocolKind() ==
+      KnownDerivableProtocolKind::Decodable) {
+    if (varType->isEqual(DC->getDeclaredInterfaceType())) {
```

Comment:
I'm afraid this fix is not general enough. You might trigger a similar circularity condition with a more complex cycle, for example involving multiple classes where each one has a stored property initialized with an expression that constructs the other.

Is there a way to fix the underlying issue with setting type witnesses re-entrantly?

Reply:
Yeah, you’re are right. There’s probably a way to solve it more generally, I will have to think more about it. I initially thought I could just check if we’ve recorded an invalid witness before (or store that information somewhere) and use that to break the cycle but I’m not sure if that’s the right solution.

Reply:
I think it's something like this:

```
- begin resolving witness for foo
  - get the interface type of something, or do something else triggering name lookup
    - name lookup goes and tries to synthesize an implicit member
      - triggers a conformance lookup
        - begin resolving witness for foo
          - do some more work
          - fail, record invalid witness
 - do some more work
 - fail, record invalid witness -- assertion here?
```

Reply:
Yeah that’s pretty much it. ValueWitnessRequest triggers a conformance lookup (when we add a name lookup entry which is triggered by a lookup call in constraint system) and that triggers witness derivation which fails and we record a null witness. Then we perform the derivation again and we hit an assert when recording it. If I could check for a previously recorded witness early, then I think I could probably just return nullptr/failure to avoid triggering derivation again?

Reply:
I think a more robust fix is to turn witness derivation into a request so that we don’t attempt it again when it’s triggered recursively. This seems more robust than checking if a witness was already recorded after every call that might potentially call back into name lookup. What do you think?

Reply:
Hmm, I think that would certainly be better, but do you mean derivation logic itself or just the resolution (resolveWitnessViaDerivation) logic?

Reply:
You could even do both. Also maybe @CodaFi has some suggestions.

Reply:
Okay, I have turned `deriveProtocolRequirement` into a request, but that hasn't solved the problem yet. It seems like we don't have an active request on the stack when the next one kicks in (which seems to imply there is no recursion?). I think I might have to turn `resolveWitnessViaDerivation` into a request as well, but that's a bit tricky - I don't think I can move the logic itself into the request because I don't have access to all the dependencies, but I can create a `ConformanceChecker` inside it and have it call `resolveWitnessViaDerivation`, similar to how `ValueWitnessRequest` works.

I was also trying out the `-debug-cycles -build-request-dependency-graph` flags but I don't see the request in the graph. Am I using it wrong?

Reply:
Yes, `-frontend -typecheck -debug-cycles -build-request-dependency-graph` should highlight any cycles, and the evaluator should emit an error too, IIRC defaulting to `circular reference`. There might still be a recursion, though, if a reentrant evaluation is passed a different argument — it's all about which parameters should contribute to the identity of the request.

Reply:
Hmm, this is what the graph looks like:

```
===CYCLE DETECTED===
 `--TypeCheckSourceFileRequest(source_file "/Users/spare/Desktop/test.swift")
     `--ExpandASTScopeRequest(ASTSourceFileScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
     `--FragileFunctionKindRequest(0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift") -> none, allowUsableFromInline: false
     `--GenericParamListRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     `--GenericSignatureRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     `--HasCircularInheritanceRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 0
     |   `--SuperclassDeclRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     |       `--InheritedDeclsReferencedRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, 0)
     |       `--UnderlyingTypeDeclsReferencedRequest(Swift.(file).Codable)
     `--StoredPropertiesRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> {test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9}
     |   `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
     |   `--StorageImplInfoRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
     |       `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
     `--DirectLookupRequest(directly looking up 'deinit' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
     `--GetDestructorRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     |   `--AccessLevelRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     `--EmittedMembersRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
         `--InterfaceTypeRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
         `--SuperclassTypeRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, interface)
         |   `--InheritedTypeRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, 0, interface)
         |       `--ResolveTypeRequest(while resolving type , Codable)
         |           `--UnqualifiedLookupRequest(looking up 'Codable' from 0x7fc4ac020d20 ClassDecl name=Foo with options { TypeLookup })
         |           |   `--LookupInModuleRequest(0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift", 'Codable', UnqualifiedLookup, TypesOnly, 0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift")
         |           `--GenericParamListRequest(Swift.(file).Codable)
         |           `--InterfaceTypeRequest(Swift.(file).Codable)
         |           `--FragileFunctionKindRequest(0x7fc4ac020d20 ClassDecl name=Foo) -> none, allowUsableFromInline: false
         `--HasDefaultInitRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 1
         |   `--HasUserDefinedDesignatedInitRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 0
         |   `--AreAllStoredPropertiesDefaultInitableRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 1
         |       `--PropertyWrapperBackingPropertyInfoRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         |           `--PropertyWrapperBackingPropertyTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         |               `--AttachedPropertyWrapperTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9, 0)
         |                   `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         `--SynthesizeDefaultInitRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> test.(file).Foo.init()@/Users/spare/Desktop/test.swift:1:7
         `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--InterfaceTypeRequest(Swift.(file).Decodable)
         `--GenericSignatureRequest(Swift.(file).Decodable)
         `--SPIGroupsRequest(Swift.(file).Decodable) -> {}
         `--InterfaceTypeRequest(Swift.(file).Decodable.init(from:))
         |   `--GenericParamListRequest(Swift.(file).UnsafeMutableRawPointer)
         |   `--GenericParamListRequest(Swift.(file).UnsafeRawPointer)
         `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers, NL_IncludeAttributeImplements })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options { IncludeAttrImplements })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options { IncludeAttrImplements })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options { IncludeAttrImplements })
         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options {  })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
         `--IsFinalRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
         `--DirectLookupRequest(directly looking up 'CodingKeys' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
         `--IsImplicitlyUnwrappedOptionalRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         `--InterfaceTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
             `--NamingPatternRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
                 `--PatternBindingEntryRequest((unknown decl), 0)
                     `--PatternTypeRequest((pattern @ 0x7fc4ac020ee8)) -> _
                     `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
                     `--UnqualifiedLookupRequest(looking up 'Foo' from 0x7fc4ac020f00 Initializer PatternBinding 0x0x7fc4ac020fa8 #0 with options { AllowProtocolMembers, IncludeOuterResults })
                     |   `--ExpandASTScopeRequest(NominalTypeBodyScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
                     |   |   `--ExpandASTScopeRequest(PatternEntryDeclScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
                     |   |       `--ExpandASTScopeRequest(PatternEntryInitializerScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
                     |   `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'Foo', { NL_ProtocolMembers, NL_RemoveNonVisible, NL_RemoveOverridden, NL_KnownCascadingDependency })
                     |   |   `--DirectLookupRequest(directly looking up 'Foo' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                     |   |   `--DirectLookupRequest(directly looking up 'Foo' on Swift.(file).Encodable with options {  })
                     |   |   `--DirectLookupRequest(directly looking up 'Foo' on Swift.(file).Decodable with options {  })
                     |   `--LookupInModuleRequest(0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift", 'Foo', UnqualifiedLookup, Overloadable, 0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift")
                     `--ResolveTypeRequest(while resolving type , Foo)
                     |   `--FragileFunctionKindRequest(0x7fc4ac020f00 Initializer PatternBinding 0x0x7fc4ac020fa8 #0) -> none, allowUsableFromInline: false
                     `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
                     |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                     `--QualifiedLookupRequest(0x7fc4ac020f00 Initializer PatternBinding 0x0x7fc4ac020fa8 #0, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init', { NL_ProtocolMembers })
                     |   `--DirectLookupRequest(directly looking up 'init' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                     |   `--DirectLookupRequest(directly looking up 'init' on Swift.(file).Encodable with options {  })
                     |   `--DirectLookupRequest(directly looking up 'init' on Swift.(file).Decodable with options {  })
                     `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                     `--ValueWitnessRequest(Foo: Decodable module test, Swift.(file).Decodable.init(from:))
                         `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                         `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
                         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
                         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers, NL_IncludeAttributeImplements })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options { IncludeAttrImplements })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options { IncludeAttrImplements })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options { IncludeAttrImplements })
                         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
                         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options {  })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
                         `--DirectLookupRequest(directly looking up 'CodingKeys' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                         `--InterfaceTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9) (cyclic dependency)
===CYCLE DETECTED===
 `--TypeCheckSourceFileRequest(source_file "/Users/spare/Desktop/test.swift")
     `--ExpandASTScopeRequest(ASTSourceFileScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
     `--FragileFunctionKindRequest(0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift") -> none, allowUsableFromInline: false
     `--GenericParamListRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     `--GenericSignatureRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     `--HasCircularInheritanceRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 0
     |   `--SuperclassDeclRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     |       `--InheritedDeclsReferencedRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, 0)
     |       `--UnderlyingTypeDeclsReferencedRequest(Swift.(file).Codable)
     `--StoredPropertiesRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> {test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9}
     |   `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
     |   `--StorageImplInfoRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
     |       `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
     `--DirectLookupRequest(directly looking up 'deinit' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
     `--GetDestructorRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     |   `--AccessLevelRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
     `--EmittedMembersRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
         `--InterfaceTypeRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
         `--SuperclassTypeRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, interface)
         |   `--InheritedTypeRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, 0, interface)
         |       `--ResolveTypeRequest(while resolving type , Codable)
         |           `--UnqualifiedLookupRequest(looking up 'Codable' from 0x7fc4ac020d20 ClassDecl name=Foo with options { TypeLookup })
         |           |   `--LookupInModuleRequest(0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift", 'Codable', UnqualifiedLookup, TypesOnly, 0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift")
         |           `--GenericParamListRequest(Swift.(file).Codable)
         |           `--InterfaceTypeRequest(Swift.(file).Codable)
         |           `--FragileFunctionKindRequest(0x7fc4ac020d20 ClassDecl name=Foo) -> none, allowUsableFromInline: false
         `--HasDefaultInitRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 1
         |   `--HasUserDefinedDesignatedInitRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 0
         |   `--AreAllStoredPropertiesDefaultInitableRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> 1
         |       `--PropertyWrapperBackingPropertyInfoRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         |           `--PropertyWrapperBackingPropertyTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         |               `--AttachedPropertyWrapperTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9, 0)
         |                   `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         `--SynthesizeDefaultInitRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7) -> test.(file).Foo.init()@/Users/spare/Desktop/test.swift:1:7
         `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--InterfaceTypeRequest(Swift.(file).Decodable)
         `--GenericSignatureRequest(Swift.(file).Decodable)
         `--SPIGroupsRequest(Swift.(file).Decodable) -> {}
         `--InterfaceTypeRequest(Swift.(file).Decodable.init(from:))
         |   `--GenericParamListRequest(Swift.(file).UnsafeMutableRawPointer)
         |   `--GenericParamListRequest(Swift.(file).UnsafeRawPointer)
         `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers, NL_IncludeAttributeImplements })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options { IncludeAttrImplements })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options { IncludeAttrImplements })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options { IncludeAttrImplements })
         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options {  })
         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
         `--IsFinalRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7)
         `--DirectLookupRequest(directly looking up 'CodingKeys' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
         `--IsImplicitlyUnwrappedOptionalRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
         `--InterfaceTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
             `--NamingPatternRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
                 `--PatternBindingEntryRequest((unknown decl), 0)
                     `--PatternTypeRequest((pattern @ 0x7fc4ac020ee8)) -> _
                     `--AttachedPropertyWrappersRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9)
                     `--UnqualifiedLookupRequest(looking up 'Foo' from 0x7fc4ac020f00 Initializer PatternBinding 0x0x7fc4ac020fa8 #0 with options { AllowProtocolMembers, IncludeOuterResults })
                     |   `--ExpandASTScopeRequest(NominalTypeBodyScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
                     |   |   `--ExpandASTScopeRequest(PatternEntryDeclScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
                     |   |       `--ExpandASTScopeRequest(PatternEntryInitializerScope
, (swift::ASTSourceFileScope*) 0x7fc4ac0210a8
)
                     |   `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'Foo', { NL_ProtocolMembers, NL_RemoveNonVisible, NL_RemoveOverridden, NL_KnownCascadingDependency })
                     |   |   `--DirectLookupRequest(directly looking up 'Foo' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                     |   |   `--DirectLookupRequest(directly looking up 'Foo' on Swift.(file).Encodable with options {  })
                     |   |   `--DirectLookupRequest(directly looking up 'Foo' on Swift.(file).Decodable with options {  })
                     |   `--LookupInModuleRequest(0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift", 'Foo', UnqualifiedLookup, Overloadable, 0x7fc4ac0208d8 FileUnit file="/Users/spare/Desktop/test.swift")
                     `--ResolveTypeRequest(while resolving type , Foo)
                     |   `--FragileFunctionKindRequest(0x7fc4ac020f00 Initializer PatternBinding 0x0x7fc4ac020fa8 #0) -> none, allowUsableFromInline: false
                     `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
                     |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                     `--QualifiedLookupRequest(0x7fc4ac020f00 Initializer PatternBinding 0x0x7fc4ac020fa8 #0, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init', { NL_ProtocolMembers })
                     |   `--DirectLookupRequest(directly looking up 'init' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                     |   `--DirectLookupRequest(directly looking up 'init' on Swift.(file).Encodable with options {  })
                     |   `--DirectLookupRequest(directly looking up 'init' on Swift.(file).Decodable with options {  })
                     `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                     `--ValueWitnessRequest(Foo: Decodable module test, Swift.(file).Decodable.init(from:))
                         `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                         `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
                         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
                         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers, NL_IncludeAttributeImplements })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options { IncludeAttrImplements })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options { IncludeAttrImplements })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options { IncludeAttrImplements })
                         `--ResolveImplicitMemberRequest(test.(file).Foo@/Users/spare/Desktop/test.swift:1:7, resolve Decodable.init(from:))
                         |   `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for Foo in test.test)
                         `--QualifiedLookupRequest(0x7fc4ac020d20 ClassDecl name=Foo, {test.(file).Foo@/Users/spare/Desktop/test.swift:1:7}, 'init(from:)', { NL_ProtocolMembers })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Encodable with options {  })
                         |   `--DirectLookupRequest(directly looking up 'init(from:)' on Swift.(file).Decodable with options {  })
                         `--DirectLookupRequest(directly looking up 'CodingKeys' on test.(file).Foo@/Users/spare/Desktop/test.swift:1:7 with options {  })
                         `--InterfaceTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9) (cyclic dependency)
                         `--LookupConformanceInModuleRequest(looking up conformance to Swift.(file).Decodable for <<error type>> in test.test)
                         `--InterfaceTypeRequest(test.(file).Foo.foo@/Users/spare/Desktop/test.swift:3:9) (cyclic dependency)
```

Reply:
According to this `InterfaceTypeRequest` is a cycle dependency, but I am not sure if that has anything to do with the bug. I think what's happening is that we're setting the witness (and ending the request) before we kick off another request.

Reply:
> I don't see the request in the graph

I am pretty sure this can only mean one thing: the `InterfaceTypeRequest` cycle precedes and prevents value witness synthesis.